### PR TITLE
Get rules by id when editing branch protection rule

### PR DESCRIPTION
**Pull request by Zettat123**
When users rename an existing branch protection rule, a new rule with the new name will be created and the old rule will still exist.

image

---

image

The reason is that the `SettingsProtectedBranchPost` function only get branch protection rule by name before updating or creating a rule. When the rule name changes, the function cannot find the existing rule so it will create a new rule rather than update the existing rule. To fix the bug, the function should get rule by id first.

### CODE REVIEW

File: routers/web/repo/setting_protected_branch.go [hunk 1]
```diff
@@ -166,10 +166,12 @@ func SettingsProtectedBranchPost(ctx *context.Context) {
 	}
 
 	var err error
-	protectBranch, err = git_model.GetProtectedBranchRuleByName(ctx, ctx.Repo.Repository.ID, f.RuleName)
-	if err != nil {
-		ctx.ServerError("GetProtectBranchOfRepoByName", err)
-		return
+	if f.RuleID > 0 {
```
Comment (wolfogre) [inline routers/web/repo/setting_protected_branch.go:169]: Maybe:

```go
	if f.RuleID > 0 {
		var err error
		protectBranch, err = git_model.GetProtectedBranchRuleByID(ctx, ctx.Repo.Repository.ID, f.RuleID)
```

**Review by wolfogre — APPROVED** [review]

File: routers/web/repo/setting_protected_branch.go (hunk 1 above)
Comment (Zettat123) [reply routers/web/repo/setting_protected_branch.go:169]: You're right. I will fix it.

File: routers/web/repo/setting_protected_branch.go [hunk 2]
```diff
@@ -165,12 +165,25 @@ func SettingsProtectedBranchPost(ctx *context.Context) {
 		return
 	}
 
+	// FIXME: If a new ProtectBranch has a duplicate RuleName, an error should be returned.
+
+	// Currently, if a new ProtectBranch with the same name is created,
+	// the existing ProtectBranch will be updated.
+	// But we cannot modify this logic now because many unit tests rely on it.
+
```
Comment (wolfogre) [inline routers/web/repo/setting_protected_branch.go]: I think it should be:

- When creating a new ProtectBranch with the same name, the RuleID should be 0, so it's fine to update the existing one. (Update an existing ProtectBranch by name)
- If the RuleID isn't 0 and the RuleName is duplicate, it should return an error. (Update the name of an existing ProtectBranch by id)

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
LG-TM bot

**Pull request merged**